A. Commit message:
Prevent pattern matching abuse (CVE-2024-31228) for older version 4.0.14

B. Change summary:
Added protection against abusive patterns by limiting the recursion depth in stringmatchlen function. Adjusted function signature to include a 'nesting' parameter, which prevents excessive recursion that could lead to a crash or denial of service.

C. Compatibility Risk:
Low

D. Fixed Code:
diff --git a/src/util.c b/src/util.c
index 26d92b9..c32cbee 100644
--- a/src/util.c
+++ b/src/util.c
@@ -40,7 +40,8 @@
 int stringmatchlen(const char *pattern, int patternLen,
         const char *string, int stringLen, int nocase)
 {
-    while(patternLen && stringLen) {
+    int nesting = 0;
+    while(patternLen && stringLen && nesting <= 1000) {
         switch(pattern[0]) {
         case '*':
             while (pattern[1] == '*') {
@@ -51,7 +52,7 @@ int stringmatchlen(const char *pattern, int patternLen,
                 return 1; /* match */
             while(stringLen) {
                 if (stringmatchlen(pattern+1, patternLen-1,
-                            string, stringLen, nocase))
+                            string, stringLen, nocase) && ++nesting <= 1000)
                     return 1; /* match */
                 string++;
                 stringLen--;
@@ -189,7 +190,7 @@ int stringmatchlen(const char *pattern, int patternLen,
 int stringmatch(const char *pattern, const char *string, int nocase) {
     return stringmatchlen(pattern,strlen(pattern),string,strlen(string),nocase);
 }
-
+
diff --git a/tests/unit/keyspace.tcl b/tests/unit/keyspace.tcl
index 43690d0..b424212 100644
--- a/tests/unit/keyspace.tcl
+++ b/tests/unit/keyspace.tcl
@@ -499,4 +499,10 @@ foreach {type large} [array get largevalue] {
         r SET aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa 1
         r KEYS "a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*b"
     } {}
+
+    test {Regression for pattern matching very long nested loops} {
+        r flushdb
+        r SET [string repeat "a" 50000] 1
+        r KEYS [string repeat "*?" 50000]
+    } {}
\ No newline at end of file